File: waymark3/core/demand.py
"""Demand classes (design §10): every action's effort, computed and rendered.

Every input field is a demand the server makes of the human. The class is
derived from the declaration and rendered on the wire (``"effort"``), so
clients, agents, and checks share one vocabulary:

- ``assent``      — empty or const-only input: one click.
- ``selection``   — every field enumerable (enum / const / boolean /
                    acceptance set / picker): choosing, not typing.
- ``recall``      — open but format-constrained typing.
- ``composition`` — unbounded text; drafts are the floor here.

(``traversal`` — links and summaries — is below the action layer.)
"""
from __future__ import annotations

from typing import Any

from .actions import ActionDef

ASSENT = "assent"
SELECTION = "selection"
RECALL = "recall"
COMPOSITION = "composition"


def _widget(prop: dict[str, Any]) -> str | None:
    return (prop.get("x-display") or {}).get("widget")


def _resolve(schema: dict[str, Any], prop: dict[str, Any]) -> dict[str, Any]:
    """Follow one level of local $ref / single-element allOf (pydantic wraps
    enums this way)."""
    if len(prop.get("allOf", ())) == 1:
        prop = {**prop, **prop["allOf"][0]}
    ref = prop.get("$ref", "")
    if ref.startswith("#/$defs/"):
        target = (schema.get("$defs") or {}).get(ref.rsplit("/", 1)[-1], {})
        prop = {**target, **{k: v for k, v in prop.items() if k != "$ref"}}
    return prop


def _variants(prop: dict[str, Any]) -> list[dict[str, Any]]:
    """anyOf/oneOf branches minus null, else the property itself."""
    branches = prop.get("anyOf") or prop.get("oneOf")
    if not branches:
        return [prop]
    out = [b for b in branches if b.get("type") != "null"]
    return out or [prop]


def demand_class(defn: ActionDef, schema: dict[str, Any] | None) -> str:
    """The demand class of an action, from its rendered input schema and its
    guards' declared acceptance sets."""
    if defn.input is None or schema is None:
        return ASSENT
    props: dict[str, Any] = schema.get("properties") or {}
    if not props:
        return ASSENT

    accepted_fields: set[str] = set()
    for guard in defn.guards:
        for g in guard.iter_leaves():
            if g.accepts is not None:
                # a Relation closes all its judged fields (design §5)
                accepted_fields.update(
                    g.judges if g.is_relation else (g.judges[0],))
    key_field = defn.place.key if defn.place else None

    worst = ASSENT
    order = {ASSENT: 0, SELECTION: 1, RECALL: 2, COMPOSITION: 3}
    for name, raw in props.items():
        prop = _resolve(schema, raw)
        if "const" in prop or name == key_field:
            continue  # pre-bound: no demand
        cls = _field_class(schema, name, prop, accepted_fields)
        if order[cls] > order[worst]:
            worst = cls
    return worst


def _field_class(schema: dict[str, Any], name: str, prop: dict[str, Any],
                 accepted_fields: set[str]) -> str:
    if _widget(prop) == "prose":
        return COMPOSITION
    if "enum" in prop or "const" in prop:
        return SELECTION
    if name in accepted_fields or _widget(prop) == "resource":
        return SELECTION
    for v in _variants(prop):
        v = _resolve(schema, v)
        t = v.get("type")
        if t == "boolean" or "enum" in v or "const" in v:
            continue
        if t == "string":
            max_len = v.get("maxLength")
            if v.get("format") or (max_len is not None and max_len < 280):
                return RECALL
            return COMPOSITION if _widget(prop) == "prose" else RECALL
        if t in ("number", "integer"):
            return RECALL
        if t in ("array", "object"):
            return RECALL
    return SELECTION

File: waymark10/src/waymark10/demand.clj
(ns waymark10.demand
  "Demand classes (waymark9 core/demand.py, design §10): every input
  field is a demand the server makes of the human, and the class is
  derived from the declaration and rendered on the wire (\"effort\")
  so clients, agents, and checks share one vocabulary:

  - assent      — empty or const-only input: one click.
  - selection   — every field enumerable (enum / const / boolean /
                  acceptance set / ref picker): choosing, not typing.
  - recall      — open but format-constrained typing.
  - composition — unbounded prose; drafts are the floor here.

  (traversal — links and summaries — is below the action layer.)

  ENVELOPE-ONLY: the class is a projection of the declaration, never
  a fact of it — it rides every rendered action entry and is NOT
  fingerprinted (two laws that differ only in effort are the same
  law, because effort is derived, not declared).

  Ported deviations, recorded:
  - waymark9 excluded fields a When conditionally requires (the entry
    reflects the base branch). v10 declares no conditional-input
    grammar yet, so there is nothing to exclude — the port drops
    that branch with this note as its tombstone.
  - waymark9 resolved pydantic's $ref/allOf wrappers; v10's published
    schemas arrive inlined (schema/json-schema), so resolution is
    :oneOf/:anyOf variant walking only."
  (:require [waymark10.guards :as g]))

(set! *warn-on-reflection* true)

(def classes ["assent" "selection" "recall" "composition"])

(def ^:private order {"assent" 0 "selection" 1 "recall" 2 "composition" 3})

(defn- widget [prop] (get-in prop [:x-display :widget]))

(defn- variants
  "anyOf/oneOf branches minus null, else the property itself."
  [prop]
  (let [branches (or (:anyOf prop) (:oneOf prop))]
    (if-not (seq branches)
      [prop]
      (let [out (into [] (remove #(= "null" (:type %))) branches)]
        (if (seq out) out [prop])))))

(defn- accepted-fields
  "Input fields some guard's declared acceptance set closes: a
  single-field :accepts closes its judged field; a relation closes
  ALL its judged fields (design §5)."
  [defn']
  (into #{}
        (mapcat (fn [leaf]
                  (when (:accepts leaf)
                    (if (:relation leaf)
                      (:judges leaf)
                      [(first (:judges leaf))]))))
        (mapcat g/iter-leaves (:guards defn'))))

(defn heavier?
  "Is class a a heavier demand than class b? The order is the
  vocabulary's own: assent < selection < recall < composition. Public
  because the usability battery (waymark10.usability) asks the
  comparison out loud — a gesture may not collect more than a
  selection — and a second spelling of this order would be a second
  opinion about what effort means."
  [a b]
  (> (long (order a)) (long (order b))))

(defn field-class
  "The demand class of ONE input field: fname its key, prop its
  RENDERED JSON-Schema property (keyword keys), accepted the set of
  fields some guard's declared acceptance set closes. Public for the
  same reason `heavier?` is — the usability battery reads effort
  per FIELD, not per action, because the fix is per field."
  [fname prop accepted]
  (cond
    (= "prose" (widget prop)) "composition"
    (or (contains? prop :enum) (contains? prop :const)) "selection"
    (or (contains? accepted fname)
        (= "resource" (widget prop))
        (:x-ref prop)
        (= "waymark-ref" (:format prop))) "selection"
    :else
    (or (some (fn [v]
                (let [t (:type v)]
                  (cond
                    (or (= "boolean" t)
                        (contains? v :enum) (contains? v :const)) nil
                    (= "string" t) (if (= "prose" (widget prop))
                                     "composition" "recall")
                    (#{"number" "integer" "array" "object"} t) "recall"
                    :else nil)))
              (variants prop))
        "selection")))

(defn effort
  "The demand class of one action, from its RENDERED input schema
  (acceptance sets already folded as enums — keyword keys, before the
  wire boundary) and its guards' declared acceptance sets. key-field
  is the place-scope key when the action is placed: pre-bound by the
  part binding (or const already), it demands nothing."
  [defn' input-schema & [key-field]]
  (let [props (:properties input-schema)]
    (if (or (nil? (:input defn')) (empty? props))
      "assent"
      (let [accepted (accepted-fields defn')]
        (reduce
         (fn [worst [fname prop]]
           (if (or (contains? prop :const) (= fname key-field))
             worst                        ; pre-bound: no demand
             (let [cls (field-class fname prop accepted)]
               (if (> (order cls) (order worst)) cls worst))))
         "assent"
         props)))))
